SLING-10123 Distribution agent queue processor should implement a backoff in case of retries for processing an item

mohiaror opened the pull request:
Adding linear backoff for a random amount of time between 0 and 30 seconds while retrying to distribute an item in case of recoverable distribution exception.

mohiaror committed:
SLING-10123 Distribution agent queue processor should implement a backoff in case of retries for processing an item

Adding linear backoff for a random amount of time between 0 and 30 seconds while retrying to distribute an item in case of recoverable distribution exception.

actinium15 commented on src/main/java/org/apache/sling/distribution/agent/impl/SimpleDistributionAgentQueueProcessor.java:
```diff
@@ -173,6 +176,16 @@ private boolean processQueueItem(String queueName, DistributionQueueEntry queueE
+        int random = (int)(30 * Math.random() + 0);
+        try {
+            Thread.sleep(random * 1000);
+        } catch (InterruptedException e) {
```
nit: conventionally, ignored execeptions should be named differently
```suggestion
        } catch (InterruptedException ign) {
```

actinium15 commented on src/main/java/org/apache/sling/distribution/agent/impl/SimpleDistributionAgentQueueProcessor.java:
```diff
@@ -173,6 +176,16 @@ private boolean processQueueItem(String queueName, DistributionQueueEntry queueE
     }
 
+    private void addRandomDelay(){
+        int random = (int)(30 * Math.random() + 0);
```
I think we need to penalize retries progressively, but in a probabilistic manner - so I'd propose:
* either, `rand(0, min(retry_count, 30)]` (pick a random number between 1 and `retry_count` to decide how long to wait)
* or, `min(rand[max(0, retry_count), 30], 30)` (pick a random number between `retry_count` and 30, capped at 30)

the latter has the characteristic that after 30 reattempts the wait will be a constant 30s (and thus different clients pointing to the same service may start to work in lock-step), whereas the former will always have variance.

The latter penalizes strongly on retries, the former is more lenient.

mohiaror replied:
@actinium15 I have updated my PR to implement the first approach suggested by you.

mohiaror committed:
SLING-10123 Distribution agent queue processor should implement a backoff in case of retries for processing an item

Incorporating review feedback.

actinium15 commented on src/main/java/org/apache/sling/distribution/agent/impl/SimpleDistributionAgentQueueProcessor.java:
```diff
@@ -147,6 +147,9 @@ private boolean processQueueItem(String queueName, DistributionQueueEntry queueE
                     distributionLog.debug("could not deliver package {}", distributionPackage.getId(), e);
+                    // since there is a recoverable error, it is possible that the same error is observed on the retry
+                    // we should add a linear backoff using random delay before re-attempting to distribute the same item.
+                    addRandomDelay(queueItemStatus.getAttempts());
```
Thinking more on this, we'd want to add this to the top of the `try` block - something like:
```
try {
    int processingAttempts = queueItemStatus.getAttempts();
    if (processingAttempts > 0) {
        // this is not the first attempt - throttle
        addRandomDelay(queueItemStatus.getAttempts());
    }
    ... // business as usual
} catch (...) {
...
}
```

I say so because even for non-recoverable exceptions (codified as `catch (Throwable e) {...}`) we should be adding the delay.

mohiaror replied:
For non-recoverable exceptions there will be no retries AFAIK. Such items will directly be moved to the error queue.

actinium15 replied:
hm, the last two scenarios we faced (which made us re-prio this issue) weren't moved to error-queue (and would only do so now because we patched them upstream).
Essentially, there are cases where we can't rely on the `catch (Throwable e) {}` clause in this `try {}`, or the `if (... != null) { }` prior to it (e.g., an unchecked exception thrown by `DistributionUtils.getResourceResolver` or `distributionPackageExporter.getPackage`). Which is where adding the delay suggested [here](https://github.com/apache/sling-org-apache-sling-distribution-core/pull/49/files#r593188382) would help.
Also, from a semantic PoV, doing so would mean "we don't care _what_ caused the queue-item to have been retried, but since we _are_ seeing a retry, let's throttle it" - which aiu is what we want.

Do you see any disadvantage if we were to move the `addRandomDelay` call to the top of the `try` block?

mohiaror replied:
@actinium15 I thought you were suggesting to move this at the top of the inner try block (because of the catch block you added in the example code). The inner try block only tries to distribute a valid distribution package. However, we can add this delay (based on greater than zero retries) at the top of the outer try block. I will update the PR accordingly.

actinium15 commented on src/main/java/org/apache/sling/distribution/agent/impl/SimpleDistributionAgentQueueProcessor.java:
```diff
@@ -173,6 +176,18 @@ private boolean processQueueItem(String queueName, DistributionQueueEntry queueE
+    private void addRandomDelay(int retryAttempts){
+        int min = 1;
+        int max = Math.min(retryAttempts, 30);
+        int random = (int)(Math.random() *  (max - min + 1) + min);
```
thanks! this looks good to me.

mohiaror committed:
SLING-10123 Distribution agent queue processor should implement a backoff in case of retries for processing an item

We want to add the backoff for any retry of distribution item and not just in case of recoverable exception.

actinium15 reviewed (approved):
Thanks for incorporating my suggestions...the changeset looks good to me!
